sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedReadFromBoundedSource.java

```diff
@@ -0,0 +1,320 @@
+        this.residualElements = residualElements;
+        this.residualSource = residualSource;
+        this.options = Preconditions.checkNotNull(options, "options");
+        this.currentElementInList = true;
+        this.currentIndexInList = -1;
+        this.reader = null;
+        this.done = false;
+      }
+
+      @Override
+      public boolean start() throws IOException {
+        return advance();
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        if (advanceInList()) {
+          done = false;
+        } else {
+          done = !advanceInReader();
+        }
+        return !done;
+      }
+
+      private boolean advanceInList() {
+        if (residualElements != null && currentIndexInList + 1 < residualElements.size()) {
+          ++currentIndexInList;
+          currentElementInList = true;
+          return true;
+        } else {
+          currentElementInList = false;
+          return false;
+        }
+      }
+
+      private boolean advanceInReader() throws IOException {
+        boolean hasNext;
+        if (boundedSource == null) {
+          hasNext = false;
+        } else if (reader == null) {
+          reader = residualSource.createReader(options);
+          hasNext = reader.start();
+        } else {
+          hasNext = reader.advance();
+        }
+        return hasNext;
+      }
+
+      @Override
+      public void close() throws IOException {
+        reader.close();
+      }
+
+      @Override
+      public T getCurrent() throws NoSuchElementException {
+        if (currentElementInList) {
+          return residualElements.get(currentIndexInList).getValue();
+        } else {
+          return reader.getCurrent();
+        }
+      }
+
+      @Override
+      public Instant getCurrentTimestamp() throws NoSuchElementException {
+        if (currentElementInList) {
+          return residualElements.get(currentIndexInList).getTimestamp();
+        } else {
+          return reader.getCurrentTimestamp();
+        }
+      }
+
+      @Override
+      public Instant getWatermark() {
+        return done ? BoundedWindow.TIMESTAMP_MAX_VALUE : BoundedWindow.TIMESTAMP_MIN_VALUE;
+      }
+
+      @Override
+      public Checkpoint<T> getCheckpointMark() {
+        List<TimestampedValue<T>> newResidualElements = Lists.newArrayList();
+        BoundedSource<T> newResidualSource;
+        if (currentElementInList) {
+          while (advanceInList()) {
+            newResidualElements.add(residualElements.get(currentIndexInList));
+          }
+          newResidualSource = residualSource;
+        } else {
+          BoundedSource<T> residualSplit = reader.splitAtFraction(reader.getFractionConsumed());
+          try {
+            while (advanceInReader()) {
+              newResidualElements.add(
+                  TimestampedValue.of(reader.getCurrent(), reader.getCurrentTimestamp()));
+            }
+          } catch (NoSuchElementException | IOException e) {
```

Comment:
Don't catch the NoSuchElementException, as it's already a runtime exception (and represents a bug in the runner implementation)